README.md
# js-base64-decode

通过 indexDB 和 webworker 优化 base64 decode 速度，适用于频繁调用 base64 decode 场景

## install

```bash
# yarn
yarn add js-base64-decode
# npm
npm i js-base64-decode --save
```

## 使用

```typescript
import { Base64 } from 'js-base64-decode'
const base64 = new Base64({ json: true })

base64.decode('some code').then((r) => {
  console.log(r)
})
```

## types

```typescript
import { IndexDBStorageOption } from 'indexdb-storage'
export interface Base64Option {
  /** webdb 名称 */
  webdb?: IndexDBStorageOption
  /** 是否需要转 json */
  json?: boolean
}
export declare type Base64Property = Required<Base64Option>
export declare class Base64 {
  /** webdb 名称 */
  private webdb
  /** 是否需要转 json */
  private json
  /** indexdb */
  private storage?
  /** web worker */
  private worker?
  constructor(option: Base64Option)
  decode<R = any>(ctx: string): Promise<R>
}
```


history.md
# 版本变更

## 0.2.3 (2021-09-17)

- feat: 把 js-base64 所有 内容导出

## 0.2.1 (2021-03-18)

- feat: 把 js-base64 decode, isVaid export 出去

## 0.2.0 (2021-03-18)

- feat: 新增 isValid

## 0.1.2 (2020-12-28)

- feat: 依赖调整


test/case/base/webpack.config.js
'use strict'
const path = require('path')
const fs = require('fs')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const extFs = require('yyl-fs')
const UglifyjsWebpackPlugin = require('uglifyjs-webpack-plugin')
const OptimizeCSSAssetsPlugin = require('optimize-css-assets-webpack-plugin')
const MiniCssExtractPlugin = require('mini-css-extract-plugin')
const webpack = require('webpack')
const extOs = require('yyl-os')

const util = require('yyl-util')

// + setting
const config = {
  proxy: {
    port: 5000,
    homePage: 'http://127.0.0.1:5000/'
  },
  alias: {
    'dirname': __dirname,
    'root': path.join(__dirname, 'dist'),
    'srcRoot': path.join(__dirname, 'src'),

    'jsDest': path.join(__dirname, 'dist/assets/js'),
    'htmlDest': path.join(__dirname, 'dist/'),
    'cssDest': path.join(__dirname, 'dist/assets/css'),
    'imagesDest': path.join(__dirname, 'dist/assets/images'),
    '~r': path.join(__dirname, '../../../output')
  },
  dest: {
    basePath: '/'
  },
  concat: {}
}

config.concat[path.join(config.alias.jsDest, 'vendors.js')] = [
  path.join(config.alias.srcRoot, 'js/a.js'),
  path.join(config.alias.srcRoot, 'js/b.js')
]

config.concat[path.join(config.alias.jsDest, 'vendorsV2.js')] = [
  path.join(config.alias.srcRoot, 'js/a.js'),
  path.join(config.alias.srcRoot, 'js/b.js'),
  path.join(config.alias.jsDest, 'index.js')
]

config.concat[path.join(config.alias.cssDest, 'vendors.css')] = [
  path.join(config.alias.srcRoot, 'css/a.css'),
  path.join(config.alias.srcRoot, 'css/b.css')
]

config.concat[path.join(config.alias.cssDest, 'vendorsV2.css')] = [
  path.join(config.alias.srcRoot, 'css/a.css'),
  path.join(config.alias.srcRoot, 'css/b.css'),
  path.join(config.alias.cssDest, 'index.css')
]

// - setting

const wConfig = {
  mode: 'development',
  entry: (() => {
    const iSrcRoot = path.isAbsolute(config.alias.srcRoot)
      ? config.alias.srcRoot
      : path.join(__dirname, config.alias.srcRoot)

    let r = {}

    // multi entry
    var entryPath = path.join(iSrcRoot, 'entry')

    if (fs.existsSync(entryPath)) {
      var fileList = extFs.readFilesSync(entryPath, /\.(jsx?|tsx?)$/)
      fileList.forEach((str) => {
        var key = path.basename(str).replace(/\.[^.]+$/, '')
        if (key) {
          r[key] = [str]
        }
      })
    }

    return r
  })(),
  output: {
    path: path.resolve(__dirname, config.alias.jsDest),
    filename: '[name]-[hash:8].js',
    chunkFilename: 'async_component/[name]-[chunkhash:8].js',
    publicPath: util.path.join(
      config.dest.basePath,
      path.relative(config.alias.root, config.alias.jsDest),
      '/'
    )
  },
  module: {
    rules: [
      {
        test: /\.jsx?$/,
        // exclude: (file) => /node_modules/.test(file) && !/\.vue\.js/.test(file),
        use: (() => {
          const loaders = [
            {
              loader: 'babel-loader',
              query: (() => {
                if (!config.babelrc) {
                  return {
                    babelrc: false,
                    cacheDirectory: true,
                    presets: ['@babel/preset-react'],
                    plugins: [
                      ['@babel/plugin-proposal-decorators', { legacy: true }],
                      ['@babel/plugin-proposal-class-properties', { loose: true }],
                      '@babel/plugin-proposal-function-sent',
                      '@babel/plugin-proposal-export-namespace-from',
                      '@babel/plugin-proposal-numeric-separator',
                      '@babel/plugin-proposal-throw-expressions',
                      '@babel/plugin-syntax-dynamic-import'
                    ]
                  }
                } else {
                  return {}
                }
              })()
            }
          ]

          return loaders
        })()
      },
      {
        test: /\.html$/,
        use: [
          {
            loader: 'html-loader'
          }
        ]
      },
      {
        test: /\.pug$/,
        oneOf: [
          {
            use: ['pug-loader']
          }
        ]
      },
      {
        test: /\.(png|jpg|gif)$/,
        use: {
          loader: 'url-loader',
          options: {
            limit: 1,
            name: '[name]-[hash:8].[ext]',
            chunkFilename: 'async_component/[name]-[chunkhash:8].js',
            outputPath: path.relative(config.alias.jsDest, config.alias.imagesDest),
            publicPath: (function () {
              let r = util.path.join(
                config.dest.basePath,
                path.relative(config.alias.root, config.alias.imagesDest),
                '/'
              )
              return r
            })()
          }
        }
      },
      {
        test: /\.css$/,
        use: [
          {
            loader: MiniCssExtractPlugin.loader,
            options: {}
          },
          {
            loader: 'css-loader'
          }
        ]
      }
    ]
  },
  resolveLoader: {
    modules: [path.join(__dirname, 'node_modules')]
  },
  resolve: {
    modules: [path.join(__dirname, 'node_modules')],
    alias: config.alias
  },
  devtool: 'source-map',
  plugins: [
    new webpack.HotModuleReplacementPlugin(),
    // 样式分离插件
    new MiniCssExtractPlugin({
      filename: util.path.join(
        path.relative(
          config.alias.jsDest,
          path.join(config.alias.cssDest, '[name]-[chunkhash:8].css')
        )
      ),
      chunkFilename: '[name]-[chunkhash:8].css',
      allChunks: true
    })
  ],
  optimization: {
    minimizer: [
      new UglifyjsWebpackPlugin({
        uglifyOptions: {
          ie8: false
        }
      }),
      new OptimizeCSSAssetsPlugin({})
    ]
  }
}

// + html output
wConfig.plugins = wConfig.plugins.concat(
  (function () {
    // html 输出
    const entryPath = util.path.join(config.alias.srcRoot, 'entry')
    let outputPath = []
    const r = []

    if (fs.existsSync(entryPath)) {
      outputPath = outputPath.concat(extFs.readFilesSync(entryPath, /(\.jade|\.pug|\.html)$/))
    }

    const outputMap = {}
    const ignoreExtName = function (iPath) {
      return iPath.replace(/(\.jade|.pug|\.html|\.js|\.css|\.ts|\.tsx|\.jsx)$/, '')
    }

    outputPath.forEach((iPath) => {
      outputMap[ignoreExtName(iPath)] = iPath
    })

    const commonChunks = []
    const pageChunkMap = {}
    Object.keys(wConfig.entry).forEach((key) => {
      let iPaths = []
      if (util.type(wConfig.entry[key]) === 'array') {
        iPaths = wConfig.entry[key]
      } else if (util.type(wConfig.entry[key]) === 'string') {
        iPaths.push(wConfig.entry[key])
      }

      let isPageModule = null
      iPaths.some((iPath) => {
        const baseName = ignoreExtName(iPath)
        if (outputMap[baseName]) {
          isPageModule = baseName
          return true
        }
        return false
      })

      if (!isPageModule) {
        commonChunks.push(key)
      } else {
        pageChunkMap[isPageModule] = key
      }
    })

    outputPath.forEach((iPath) => {
      const iBaseName = ignoreExtName(iPath)
      const iChunkName = pageChunkMap[iBaseName]
      const fileName = ignoreExtName(path.basename(iPath))
      let iChunks = []

      iChunks = iChunks.concat(commonChunks)
      if (iChunkName) {
        iChunks.push(iChunkName)
      }

      if (iChunkName) {
        const opts = {
          template: iPath,
          filename: path.relative(
            config.alias.jsDest,
            path.join(config.alias.htmlDest, `${fileName}.html`)
          ),
          chunks: iChunks,
          chunksSortMode(a, b) {
            return iChunks.indexOf(a.names[0]) - iChunks.indexOf(b.names[0])
          },
          inlineSource: '.(js|css|ts|tsx|jsx)\\?__inline$',
          minify: false
        }

        r.push(new HtmlWebpackPlugin(opts))
      }
    })

    return r
  })()
)
// - html output

// + dev server
wConfig.devServer = {
  contentBase: config.alias.root,
  injectClient: false,
  compress: true,
  port: config.proxy.port,
  hot: false,
  publicPath: config.dest.basePath,
  writeToDisk: true,
  async after() {
    if (config.proxy.homePage) {
      await extOs.openBrowser(config.proxy.homePage)
    }
  }
}
// - dev server

module.exports = wConfig


output/index.js
/*!
 * js-base64-decode cjs 0.2.3
 * (c) 2020 - 2021 jackness
 * Released under the MIT License.
 */
"use strict";Object.defineProperty(exports,"__esModule",{value:!0});var r=require("indexdb-storage");
/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */var t,e="function"==typeof atob,n="function"==typeof btoa,o="function"==typeof Buffer,u="function"==typeof TextDecoder?new TextDecoder:void 0,i="function"==typeof TextEncoder?new TextEncoder:void 0,a=[].concat("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="),c=(t={},a.forEach((function(r,e){return t[r]=e})),t),f=/^(?:[A-Za-z\d+\/]{4})*?(?:[A-Za-z\d+\/]{2}(?:==)?|[A-Za-z\d+\/]{3}=?)?$/,s=String.fromCharCode.bind(String),d="function"==typeof Uint8Array.from?Uint8Array.from.bind(Uint8Array):function(r,t){return void 0===t&&(t=function(r){return r}),new Uint8Array(Array.prototype.slice.call(r,0).map(t))},l=function(r){return r.replace(/[+\/]/g,(function(r){return"+"==r?"-":"_"})).replace(/=+$/m,"")},h=function(r){return r.replace(/[^A-Za-z0-9\+\/]/g,"")},p=function(r){for(var t,e,n,o,u="",i=r.length%3,c=0;c<r.length;){if((e=r.charCodeAt(c++))>255||(n=r.charCodeAt(c++))>255||(o=r.charCodeAt(c++))>255)throw new TypeError("invalid character found");u+=a[(t=e<<16|n<<8|o)>>18&63]+a[t>>12&63]+a[t>>6&63]+a[63&t]}return i?u.slice(0,i-3)+"===".substring(i):u},b=n?function(r){return btoa(r)}:o?function(r){return Buffer.from(r,"binary").toString("base64")}:p,y=o?function(r){return Buffer.from(r).toString("base64")}:function(r){for(var t=[],e=0,n=r.length;e<n;e+=4096)t.push(s.apply(null,r.subarray(e,e+4096)));return b(t.join(""))},A=function(r,t){return void 0===t&&(t=!1),t?l(y(r)):y(r)},x=function(r){if(r.length<2)return(t=r.charCodeAt(0))<128?r:t<2048?s(192|t>>>6)+s(128|63&t):s(224|t>>>12&15)+s(128|t>>>6&63)+s(128|63&t);var t=65536+1024*(r.charCodeAt(0)-55296)+(r.charCodeAt(1)-56320);return s(240|t>>>18&7)+s(128|t>>>12&63)+s(128|t>>>6&63)+s(128|63&t)},g=/[\uD800-\uDBFF][\uDC00-\uDFFFF]|[^\x00-\x7F]/g,v=function(r){return r.replace(g,x)},w=o?function(r){return Buffer.from(r,"utf8").toString("base64")}:i?function(r){return y(i.encode(r))}:function(r){return b(v(r))},m=function(r,t){return void 0===t&&(t=!1),t?l(w(r)):w(r)},C=function(r){return m(r,!0)},B=/[\xC0-\xDF][\x80-\xBF]|[\xE0-\xEF][\x80-\xBF]{2}|[\xF0-\xF7][\x80-\xBF]{3}/g,U=function(r){switch(r.length){case 4:var t=((7&r.charCodeAt(0))<<18|(63&r.charCodeAt(1))<<12|(63&r.charCodeAt(2))<<6|63&r.charCodeAt(3))-65536;return s(55296+(t>>>10))+s(56320+(1023&t));case 3:return s((15&r.charCodeAt(0))<<12|(63&r.charCodeAt(1))<<6|63&r.charCodeAt(2));default:return s((31&r.charCodeAt(0))<<6|63&r.charCodeAt(1))}},S=function(r){return r.replace(B,U)},E=function(r){if(r=r.replace(/\s+/g,""),!f.test(r))throw new TypeError("malformed base64.");r+="==".slice(2-(3&r.length));for(var t,e,n,o="",u=0;u<r.length;)t=c[r.charAt(u++)]<<18|c[r.charAt(u++)]<<12|(e=c[r.charAt(u++)])<<6|(n=c[r.charAt(u++)]),o+=64===e?s(t>>16&255):64===n?s(t>>16&255,t>>8&255):s(t>>16&255,t>>8&255,255&t);return o},F=e?function(r){return atob(h(r))}:o?function(r){return Buffer.from(r,"base64").toString("binary")}:E,k=o?function(r){return d(Buffer.from(r,"base64"))}:function(r){return d(F(r),(function(r){return r.charCodeAt(0)}))},T=function(r){return k(z(r))},j=o?function(r){return Buffer.from(r,"base64").toString("utf8")}:u?function(r){return u.decode(k(r))}:function(r){return S(F(r))},z=function(r){return h(r.replace(/[-_]/g,(function(r){return"-"==r?"+":"/"})))},D=function(r){return j(z(r))},R=function(r){if("string"!=typeof r)return!1;var t=r.replace(/\s+/g,"").replace(/=+$/,"");return!/[^\s0-9a-zA-Z\+/]/.test(t)||!/[^\s0-9a-zA-Z\-_]/.test(t)},Z=function(r){return{value:r,enumerable:!1,writable:!0,configurable:!0}},I=function(){var r=function(r,t){return Object.defineProperty(String.prototype,r,Z(t))};r("fromBase64",(function(){return D(this)})),r("toBase64",(function(r){return m(this,r)})),r("toBase64URI",(function(){return m(this,!0)})),r("toBase64URL",(function(){return m(this,!0)})),r("toUint8Array",(function(){return T(this)}))},L=function(){var r=function(r,t){return Object.defineProperty(Uint8Array.prototype,r,Z(t))};r("toBase64",(function(r){return A(this,r)})),r("toBase64URI",(function(){return A(this,!0)})),r("toBase64URL",(function(){return A(this,!0)}))},O="undefined"!=typeof Worker&&"undefined"!=typeof URL,P=function(){function t(t){this.webdb={name:"base64-decode"},this.json=!0;var e=t.webdb,n=t.json;void 0!==e&&(this.webdb=e),void 0!==n&&(this.json=n),this.storage=new r.IndexDBStorage({name:this.webdb.name}),O&&(this.worker=new Worker(URL.createObjectURL(new Blob(['\n"use strict";var r,t="function"==typeof atob,e="function"==typeof Buffer,n="function"==typeof TextDecoder?new TextDecoder:void 0,o=("function"==typeof TextEncoder&&new TextEncoder,[].concat("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=")),c=(r={},o.forEach((function(t,e){return r[t]=e})),r),a=/^(?:[A-Za-zd+/]{4})*?(?:[A-Za-zd+/]{2}(?:==)?|[A-Za-zd+/]{3}=?)?$/,f=String.fromCharCode.bind(String),u="function"==typeof Uint8Array.from?Uint8Array.from.bind(Uint8Array):function(r,t){return void 0===t&&(t=function(r){return r}),new Uint8Array(Array.prototype.slice.call(r,0).map(t))},i=function(r){return r.replace(/[^A-Za-z0-9+/]/g,"")},d=/[À-ß][-¿]|[à-ï][-¿]{2}|[ð-÷][-¿]{3}/g,A=function(r){switch(r.length){case 4:var t=((7&r.charCodeAt(0))<<18|(63&r.charCodeAt(1))<<12|(63&r.charCodeAt(2))<<6|63&r.charCodeAt(3))-65536;return f(55296+(t>>>10))+f(56320+(1023&t));case 3:return f((15&r.charCodeAt(0))<<12|(63&r.charCodeAt(1))<<6|63&r.charCodeAt(2));default:return f((31&r.charCodeAt(0))<<6|63&r.charCodeAt(1))}},h=t?function(r){return atob(i(r))}:e?function(r){return Buffer.from(r,"base64").toString("binary")}:function(r){if(r=r.replace(/s+/g,""),!a.test(r))throw new TypeError("malformed base64.");r+="==".slice(2-(3&r.length));for(var t,e,n,o="",u=0;u<r.length;)t=c[r.charAt(u++)]<<18|c[r.charAt(u++)]<<12|(e=c[r.charAt(u++)])<<6|(n=c[r.charAt(u++)]),o+=64===e?f(t>>16&255):64===n?f(t>>16&255,t>>8&255):f(t>>16&255,t>>8&255,255&t);return o},s=e?function(r){return u(Buffer.from(r,"base64"))}:function(r){return u(h(r),(function(r){return r.charCodeAt(0)}))},x=e?function(r){return Buffer.from(r,"base64").toString("utf8")}:n?function(r){return n.decode(s(r))}:function(r){return h(r).replace(d,A)},l=self;l.addEventListener("message",(function(r){var t;l.postMessage((t=r.data,x(i(t.replace(/[-_]/g,(function(r){return"-"==r?"+":"/"}))))))}));\n'],{type:"application/javascript"})),{name:"base64-decode"}))}return t.prototype.isValid=function(r){return R(r)},t.prototype.decode=function(r){var t,e,n,o,u,i,a;return o=this,u=void 0,a=function(){var o,u,i=this;return function(r,t){var e,n,o,u,i={label:0,sent:function(){if(1&o[0])throw o[1];return o[1]},trys:[],ops:[]};return u={next:a(0),throw:a(1),return:a(2)},"function"==typeof Symbol&&(u[Symbol.iterator]=function(){return this}),u;function a(u){return function(a){return function(u){if(e)throw new TypeError("Generator is already executing.");for(;i;)try{if(e=1,n&&(o=2&u[0]?n.return:u[0]?n.throw||((o=n.return)&&o.call(n),0):n.next)&&!(o=o.call(n,u[1])).done)return o;switch(n=0,o&&(u=[2&u[0],o.value]),u[0]){case 0:case 1:o=u;break;case 4:return i.label++,{value:u[1],done:!1};case 5:i.label++,n=u[1],u=[0];continue;case 7:u=i.ops.pop(),i.trys.pop();continue;default:if(!((o=(o=i.trys).length>0&&o[o.length-1])||6!==u[0]&&2!==u[0])){i=0;continue}if(3===u[0]&&(!o||u[1]>o[0]&&u[1]<o[3])){i.label=u[1];break}if(6===u[0]&&i.label<o[1]){i.label=o[1],o=u;break}if(o&&i.label<o[2]){i.label=o[2],i.ops.push(u);break}o[2]&&i.ops.pop(),i.trys.pop();continue}u=t.call(r,i)}catch(r){u=[6,r],n=0}finally{e=o=0}if(5&u[0])throw u[1];return{value:u[0]?u[1]:void 0,done:!0}}([u,a])}}}(this,(function(a){switch(a.label){case 0:return o=this.json,[4,null===(t=this.storage)||void 0===t?void 0:t.getItem(r)];case 1:return(u=a.sent())?[2,u]:[3,2];case 2:return this.worker?[4,new Promise((function(t,e){i.worker&&(i.worker.onmessage=function(r){t(r.data)},i.worker.postMessage(r),i.worker.onerror=function(r){e(r)})}))]:[3,5];case 3:return u=a.sent(),o&&(u=JSON.parse(u)),[4,null===(e=this.storage)||void 0===e?void 0:e.setItem(r,u)];case 4:return a.sent(),[3,7];case 5:return u=D(r),o&&(u=JSON.parse(u)),[4,null===(n=this.storage)||void 0===n?void 0:n.setItem(r,u)];case 6:a.sent(),a.label=7;case 7:return[2,u]}}))},new((i=void 0)||(i=Promise))((function(r,t){function e(r){try{c(a.next(r))}catch(r){t(r)}}function n(r){try{c(a.throw(r))}catch(r){t(r)}}function c(t){var o;t.done?r(t.value):(o=t.value,o instanceof i?o:new i((function(r){r(o)}))).then(e,n)}c((a=a.apply(o,u||[])).next())}))},t}();exports.Base64=P,exports.VERSION="3.6.0",exports.atob=F,exports.atobPolyfill=E,exports.btoa=b,exports.btoaPolyfill=p,exports.btou=S,exports.decode=D,exports.encode=m,exports.encodeURI=C,exports.encodeURL=C,exports.extendBuiltins=function(){I(),L()},exports.extendString=I,exports.extendUint8Array=L,exports.fromBase64=D,exports.fromUint8Array=A,exports.isValid=R,exports.toBase64=m,exports.toUint8Array=T,exports.utob=v,exports.version="3.6.0";


output/worker.js
"use strict";var r,t="function"==typeof atob,e="function"==typeof Buffer,n="function"==typeof TextDecoder?new TextDecoder:void 0,o=("function"==typeof TextEncoder&&new TextEncoder,[].concat("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=")),c=(r={},o.forEach((function(t,e){return r[t]=e})),r),a=/^(?:[A-Za-z\d+\/]{4})*?(?:[A-Za-z\d+\/]{2}(?:==)?|[A-Za-z\d+\/]{3}=?)?$/,f=String.fromCharCode.bind(String),u="function"==typeof Uint8Array.from?Uint8Array.from.bind(Uint8Array):function(r,t){return void 0===t&&(t=function(r){return r}),new Uint8Array(Array.prototype.slice.call(r,0).map(t))},i=function(r){return r.replace(/[^A-Za-z0-9\+\/]/g,"")},d=/[\xC0-\xDF][\x80-\xBF]|[\xE0-\xEF][\x80-\xBF]{2}|[\xF0-\xF7][\x80-\xBF]{3}/g,A=function(r){switch(r.length){case 4:var t=((7&r.charCodeAt(0))<<18|(63&r.charCodeAt(1))<<12|(63&r.charCodeAt(2))<<6|63&r.charCodeAt(3))-65536;return f(55296+(t>>>10))+f(56320+(1023&t));case 3:return f((15&r.charCodeAt(0))<<12|(63&r.charCodeAt(1))<<6|63&r.charCodeAt(2));default:return f((31&r.charCodeAt(0))<<6|63&r.charCodeAt(1))}},h=t?function(r){return atob(i(r))}:e?function(r){return Buffer.from(r,"base64").toString("binary")}:function(r){if(r=r.replace(/\s+/g,""),!a.test(r))throw new TypeError("malformed base64.");r+="==".slice(2-(3&r.length));for(var t,e,n,o="",u=0;u<r.length;)t=c[r.charAt(u++)]<<18|c[r.charAt(u++)]<<12|(e=c[r.charAt(u++)])<<6|(n=c[r.charAt(u++)]),o+=64===e?f(t>>16&255):64===n?f(t>>16&255,t>>8&255):f(t>>16&255,t>>8&255,255&t);return o},s=e?function(r){return u(Buffer.from(r,"base64"))}:function(r){return u(h(r),(function(r){return r.charCodeAt(0)}))},x=e?function(r){return Buffer.from(r,"base64").toString("utf8")}:n?function(r){return n.decode(s(r))}:function(r){return h(r).replace(d,A)},l=self;l.addEventListener("message",(function(r){var t;l.postMessage((t=r.data,x(i(t.replace(/[-_]/g,(function(r){return"-"==r?"+":"/"}))))))}));


rollup.config.js
import pkg from './package.json'
import typescript from 'rollup-plugin-typescript2'
import commonjs from 'rollup-plugin-commonjs'
import nodeResolve from 'rollup-plugin-node-resolve'
import json from 'rollup-plugin-json'
import { terser } from 'rollup-plugin-terser'
import buble from '@rollup/plugin-buble'

function buildBanner(type) {
  return [
    '/*!',
    ` * ${pkg.name} ${type} ${pkg.version}`,
    ` * (c) 2020 - ${new Date().getFullYear()} jackness`,
    ' * Released under the MIT License.',
    ' */'
  ].join('\n')
}

const IS_PUBLISH = process.env.NODE_ENV === 'production'

const config = {
  input: './src/index.ts',
  output: [],
  plugins: [
    // replace({
    //   __workerScript__: () => {
    //     // return 'sb'
    //     return fs.readFileSync('./output/index.js').toString()
    //   }
    // }),
    nodeResolve({ jsnext: true }),
    commonjs({
      include: 'node_modules/**'
    }),
    json(),
    typescript({
      typescript: require('typescript')
    }),
    buble()
  ],
  external: ['indexdb-storage']
}

export default [
  {
    input: './src/worker.ts',
    output: [
      {
        file: './output/worker.js',
        format: 'commonjs',
        exports: 'named',
        sourcemap: false
      }
    ],
    plugins: config.plugins.concat([
      terser({
        compress: {
          passes: 2
        }
      })
    ])
  },
  {
    input: config.input,
    output: [
      {
        file: './output/index.js',
        format: 'cjs',
        banner: buildBanner('cjs'),
        exports: 'named',
        sourcemap: false
      }
    ],
    plugins: config.plugins.concat([
      IS_PUBLISH &&
        terser({
          compress: {
            passes: 2
          }
        })
    ]),
    external: config.external
  },
  {
    input: config.input,
    output: [
      {
        file: './output/index.esm.js',
        format: 'esm',
        banner: buildBanner('esm'),
        sourcemap: false
      }
    ],
    plugins: config.plugins,
    external: config.external
  }
]


src/index.ts
/* eslint-disable no-useless-escape */
import { decode, isValid } from 'js-base64'
import { IndexDBStorage, IndexDBStorageOption } from 'indexdb-storage'
export * from 'js-base64'

const workerString = `
"use strict";var r,t="function"==typeof atob,e="function"==typeof Buffer,n="function"==typeof TextDecoder?new TextDecoder:void 0,o=("function"==typeof TextEncoder&&new TextEncoder,[].concat("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=")),c=(r={},o.forEach((function(t,e){return r[t]=e})),r),a=/^(?:[A-Za-z\d+\/]{4})*?(?:[A-Za-z\d+\/]{2}(?:==)?|[A-Za-z\d+\/]{3}=?)?$/,f=String.fromCharCode.bind(String),u="function"==typeof Uint8Array.from?Uint8Array.from.bind(Uint8Array):function(r,t){return void 0===t&&(t=function(r){return r}),new Uint8Array(Array.prototype.slice.call(r,0).map(t))},i=function(r){return r.replace(/[^A-Za-z0-9\+\/]/g,"")},d=/[\xC0-\xDF][\x80-\xBF]|[\xE0-\xEF][\x80-\xBF]{2}|[\xF0-\xF7][\x80-\xBF]{3}/g,A=function(r){switch(r.length){case 4:var t=((7&r.charCodeAt(0))<<18|(63&r.charCodeAt(1))<<12|(63&r.charCodeAt(2))<<6|63&r.charCodeAt(3))-65536;return f(55296+(t>>>10))+f(56320+(1023&t));case 3:return f((15&r.charCodeAt(0))<<12|(63&r.charCodeAt(1))<<6|63&r.charCodeAt(2));default:return f((31&r.charCodeAt(0))<<6|63&r.charCodeAt(1))}},h=t?function(r){return atob(i(r))}:e?function(r){return Buffer.from(r,"base64").toString("binary")}:function(r){if(r=r.replace(/\s+/g,""),!a.test(r))throw new TypeError("malformed base64.");r+="==".slice(2-(3&r.length));for(var t,e,n,o="",u=0;u<r.length;)t=c[r.charAt(u++)]<<18|c[r.charAt(u++)]<<12|(e=c[r.charAt(u++)])<<6|(n=c[r.charAt(u++)]),o+=64===e?f(t>>16&255):64===n?f(t>>16&255,t>>8&255):f(t>>16&255,t>>8&255,255&t);return o},s=e?function(r){return u(Buffer.from(r,"base64"))}:function(r){return u(h(r),(function(r){return r.charCodeAt(0)}))},x=e?function(r){return Buffer.from(r,"base64").toString("utf8")}:n?function(r){return n.decode(s(r))}:function(r){return h(r).replace(d,A)},l=self;l.addEventListener("message",(function(r){var t;l.postMessage((t=r.data,x(i(t.replace(/[-_]/g,(function(r){return"-"==r?"+":"/"}))))))}));
`
export interface Base64Option {
  /** webdb 名称 */
  webdb?: IndexDBStorageOption
  /** 是否需要转 json */
  json?: boolean
}

export type Base64Property = Required<Base64Option>

const WOEKER_SUPPORTED = typeof Worker !== 'undefined' && typeof URL !== 'undefined'

export class Base64 {
  /** webdb 名称 */
  private webdb: Base64Property['webdb'] = { name: 'base64-decode' }
  /** 是否需要转 json */
  private json: Base64Property['json'] = true
  /** indexdb */
  private storage?: IndexDBStorage
  /** web worker */
  private worker?: Worker
  constructor(option: Base64Option) {
    const { webdb, json } = option
    if (webdb !== undefined) {
      this.webdb = webdb
    }
    if (json !== undefined) {
      this.json = json
    }
    this.storage = new IndexDBStorage({ name: this.webdb.name })

    if (WOEKER_SUPPORTED) {
      this.worker = new Worker(
        URL.createObjectURL(new Blob([workerString], { type: 'application/javascript' })),
        {
          name: 'base64-decode'
        }
      )
    }
  }

  isValid(ctx: any) {
    return isValid(ctx)
  }

  async decode<R = any>(ctx: string): Promise<R> {
    const { json } = this
    let r = await this.storage?.getItem(ctx)
    if (r) {
      return r as R
    } else {
      if (this.worker) {
        r = await new Promise((resolve, reject) => {
          if (this.worker) {
            this.worker.onmessage = (e) => {
              resolve(e.data)
            }
            this.worker.postMessage(ctx)
            this.worker.onerror = (er) => {
              reject(er)
            }
          }
        })
        if (json) {
          r = JSON.parse(r)
        }

        await this.storage?.setItem(ctx, r)
      } else {
        r = decode(ctx)
        if (json) {
          r = JSON.parse(r)
        }
        await this.storage?.setItem(ctx, r)
      }
      return r as R
    }
  }
}
